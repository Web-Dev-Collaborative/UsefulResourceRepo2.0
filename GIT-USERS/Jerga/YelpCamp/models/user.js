var mongoose = require("mongoose");
var passportLocalMongoose = require("passport-local-mongoose");

var UserSchema = new mongoose.Schema({
    username: String,
    password: String,
    role    : String,
    rankedCamps: [{
            type: mongoose.Schema.Types.ObjectId,
            ref: "Campground"
        }],
    createdCamps: [{
            type: mongoose.Schema.Types.ObjectId,
            ref: "Campground"
    }]
});

UserSchema.plugin(passportLocalMongoose); // to login ,register....

module.exports = mongoose.model("User",UserSchema);