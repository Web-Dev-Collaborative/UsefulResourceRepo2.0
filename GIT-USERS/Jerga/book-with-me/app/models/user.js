import DS from "ember-data";

export default DS.Model.extend({
  username: DS.attr(),
  email: DS.attr(),
  password: DS.attr(),
  password_confirmation: DS.attr(),
  rentals: DS.hasMany("rental"),
  bookings: DS.hasMany("booking"),
});
