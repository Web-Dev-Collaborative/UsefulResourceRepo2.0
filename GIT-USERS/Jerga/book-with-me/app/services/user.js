import Service from "@ember/service";
import { inject as service } from "@ember/service";

export default Service.extend({
  store: service(),
  session: service("session"),
  currentUser: null,

  initUser() {
    const currentUser = this.get("currentUser");

    if (!currentUser) {
      this.get("session.store")
        .restore()
        .then(({ authenticated }) => {
          const parsedToken = this.parseJwt(authenticated.token);

          return this.createUser(parsedToken);
        });
    }
  },

  createUser(token) {
    let user = {};

    if (token.username && token.email) {
      user = this.get("store").peekRecord("user", token.userId);

      if (!user) {
        user = this.get("store").createRecord("user", {
          id: token.userId,
          username: token.username,
          email: token.email,
        });
      }

      this.set("currentUser", user);
    }

    return user;
  },

  getUser() {
    return this.get("currentUser");
  },

  parseJwt(token) {
    if (token) {
      const base64Url = token.split(".")[1];
      const base64 = base64Url.replace("-", "+").replace("_", "/");

      return JSON.parse(window.atob(base64));
    }

    return {};
  },
});
