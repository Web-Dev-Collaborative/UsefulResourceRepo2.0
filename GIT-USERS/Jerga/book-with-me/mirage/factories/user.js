import { Factory } from "ember-cli-mirage";
import bcrypt from "npm:bcryptjs";

export default Factory.extend({
  id(i) {
    return i;
  },
  username: "Test User",
  email: "test@gmail.com",
  password() {
    const salt = bcrypt.genSaltSync(10);
    return bcrypt.hashSync("testtest", salt);
  },
});
