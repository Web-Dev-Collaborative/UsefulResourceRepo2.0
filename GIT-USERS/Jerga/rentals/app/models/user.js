import DS from 'ember-data';

export default DS.Model.extend({
  name: DS.attr(),
  email: DS.attr(),
  password: DS.attr(),
  password_confirmation: DS.attr(),
  rentals: DS.hasMany('rental', { async: true }),
  bookings: DS.hasMany('bookings')
});
