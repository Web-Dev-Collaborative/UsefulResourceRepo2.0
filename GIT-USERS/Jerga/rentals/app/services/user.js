import Service from '@ember/service';
const { service } = Ember.inject;

export default Service.extend({
  user: null,
  session: service('session'),
  jwt: service('jwt'),
  storage: service('storage'),
  store: service(),

  init() {
    this.get('session.store').restore().then(({authenticated}) => {
      const parsedToken = this.get('jwt').parseJwt(authenticated.token);
      const newUser = this.createUser(parsedToken);

      this.set('user', newUser);
    });
  },

  createUser(token) {
    if (token.name && token.email) {
      return this.get('store').createRecord('user', {name: token.name, email: token.email});
    }

    return {};
  },

  isAuthenticatedUser(rentalUser) {
    return user.email === rentalUser.email;
  }
});
