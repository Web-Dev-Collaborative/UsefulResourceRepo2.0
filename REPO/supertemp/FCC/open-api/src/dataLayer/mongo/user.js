import validator from 'validator';
import UserModel from '../model/user.js';
import debug from 'debug';
import uuid from 'uuid/v4';
import { isEmpty, isString } from 'lodash';

import { asyncErrorHandler } from '../../utils';
import { verifyWebToken, namespace, updateAppMetaData } from '../../auth';

const log = debug('fcc:dataLayer:mongo:user');

function doesExist(Model, options) {
  return Model.find(options).exec();
}

export function getUser(root, { email }) {
  return new Promise(async function getUserPromise(resolve, reject) {
    if (!isString(email) || !validator.isEmail(email)) {
      reject(
        new TypeError(`Expected a valid email, got ${JSON.stringify(email)}`)
      );
      return null;
    }
    log(`finding user for ${email}`);
    const found = await UserModel.findOne({ email }).exec();
    log(`found? ${!!found}`);
    if (isEmpty(found)) {
      reject(new Error(`No user found for ${email}`));
      return null;
    }
    return resolve(found);
  });
}

export async function createUser(root, vars, ctx) {
  const { decoded } = verifyWebToken(ctx);
  const { email, name, sub: id } = decoded;
  if (!isString(email) || !validator.isEmail(email)) {
    throw new Error('You must provide a valid email');
  }
  const newUser = { name, email };
  let accountLinkId = decoded[namespace + 'accountLinkId'];
  if (accountLinkId) {
    newUser.accountLinkId = accountLinkId;
  } else {
    accountLinkId = uuid();
    newUser.accountLinkId = accountLinkId;
    updateAppMetaData(id, { accountLinkId });
  }
  const exists = await asyncErrorHandler(
    doesExist(UserModel, { accountLinkId })
  );
  if (isEmpty(exists)) {
    const newAccount = new UserModel(newUser);
    return await asyncErrorHandler(
      newAccount.save(),
      'Something went wrong creating your account, please try again'
    );
  } else {
    throw new Error('Account already in use');
  }
}

export async function deleteUser(root, vars, ctx) {
  const { decoded } = verifyWebToken(ctx);
  const { accountLinkId } = vars;
  const loggedInId = decoded[namespace + 'accountLinkId'];

  if (loggedInId !== accountLinkId) {
    throw new Error('You can delete only your account');
  }
  const removedUser = await UserModel.findOneAndRemove({ accountLinkId });
  if (!removedUser) {
    throw new Error(
      'There is no account with this accountLinkId ' + accountLinkId
    );
  }
  return removedUser;
}
