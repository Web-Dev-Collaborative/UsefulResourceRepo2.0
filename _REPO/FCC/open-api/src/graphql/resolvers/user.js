import { createUser, getUser, deleteUser } from '../../dataLayer/mongo/user';

export const userResolvers = {
  Query: {
    getUser
  },
  Mutation: {
    createUser,
    deleteUser
  }
};
