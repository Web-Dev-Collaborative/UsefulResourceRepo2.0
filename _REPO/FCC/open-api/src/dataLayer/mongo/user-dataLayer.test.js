/* global expect beforeAll afterAll */
import { createUser, getUser, deleteUser } from './user';
import UserModel from '../model/user.js';
import { isObject, isEmpty } from 'lodash';
import mongoose from 'mongoose';

const validContextForCharlie = global.mockedContextWithValidTokenForCharlie;
const validContextForLola = global.mockedContextWithValidTokenForLola;
const validContextForJane = global.mockedContextWithValidTokenForJane;

beforeAll(async function beforeAllTests() {
  await mongoose.connect(global.__MONGO_URI__);
});

afterAll(async function afterAllTests() {
  await mongoose.disconnect();
});

describe('createUser', () => {
  it('should return a User object', done => {
    expect.assertions(2);
    createUser({}, {}, validContextForLola)
      .then(result => {
        const { name, email, accountLinkId } = result;
        // there is some weird Promise thing going on with `result`
        // which screws with lodash.has()
        const hasKeys =
          !isEmpty(name) && !isEmpty(email) && !isEmpty(accountLinkId);

        expect(isObject(result)).toBe(true);
        expect(hasKeys).toBe(true);
        return;
      })
      .then(done)
      .catch(done);
  });

  it('should throw if accountLinkId is already in db', done => {
    expect.assertions(1);
    createUser({}, {}, validContextForLola).catch(err => {
      expect(err).toMatchSnapshot();
      done();
      return;
    });
  });
});

describe('getUser', () => {
  it('should return a User object fo a valid request', done => {
    expect.assertions(2);
    const email = 'lola@cbbc.tv';
    getUser({}, { email }, validContextForLola)
      .then(result => {
        const { name, email, accountLinkId } = result;
        // there is some weird Promise thing going on with `result`
        // which screws with lodash.has()
        const hasKeys =
          !isEmpty(name) && !isEmpty(email) && !isEmpty(accountLinkId);

        expect(isObject(result)).toBe(true);
        expect(hasKeys).toBe(true);
        return;
      })
      .then(done)
      .catch(done);
  });

  it('should throw for a user not found', done => {
    expect.assertions(1);
    const email = 'not@inthe.db';
    getUser({}, { email }, validContextForCharlie).catch(err => {
      expect(err.message).toContain('No user found for ');
      done();
    });
  });

  it('should throw if the supplied email is not valid', done => {
    expect.assertions(15);
    Promise.all([
      getUser({}, { email: 1 }, validContextForCharlie).catch(err => {
        expect(err).toBeInstanceOf(TypeError);
        expect(err.message).toContain('Expected a valid email');
        expect(err.message).toContain('1');
      }),
      getUser({}, { email: 'not an email' }, validContextForCharlie).catch(
        err => {
          expect(err).toBeInstanceOf(TypeError);
          expect(err.message).toContain('Expected a valid email');
          expect(err.message).toContain('not an email');
        }
      ),
      getUser({}, { email: ['nope'] }, validContextForCharlie).catch(err => {
        expect(err).toBeInstanceOf(TypeError);
        expect(err.message).toContain('Expected a valid email');
        expect(err.message).toContain('["nope"]');
      }),
      getUser({}, { email: { email: false } }, validContextForCharlie).catch(
        err => {
          expect(err).toBeInstanceOf(TypeError);
          expect(err.message).toContain('Expected a valid email');
          expect(err.message).toContain('false');
        }
      ),
      getUser({}, { email: 1 }, validContextForCharlie).catch(err => {
        expect(err).toBeInstanceOf(TypeError);
        expect(err.message).toContain('Expected a valid email');
        expect(err.message).toContain('1');
      })
    ]).then(() => {
      done();
    });
  });
});

describe('deleteUser', () => {
  it('should delete an existing user', async done => {
    const result = await createUser({}, {}, validContextForJane);
    const { accountLinkId } = result;
    const response = await deleteUser(
      {},
      { accountLinkId },
      validContextForJane
    );
    expect(response).toBeTruthy();
    expect(response.accountLinkId).toMatch(accountLinkId);
    const searchResult = await UserModel.findOne({ accountLinkId });
    expect(searchResult).toBe(null);
    done();
  });
  it('should return with an error for a non existing user', async done => {
    try {
      await deleteUser(
        {},
        { accountLinkId: global.idOfLola },
        validContextForLola
      );
    } catch (err) {
      expect(err).toMatchSnapshot();
    }
    done();
  });
  it('should refuse deletion of other users', async done => {
    try {
      await deleteUser(
        {},
        { accountLinkId: global.idOfLola },
        validContextForCharlie
      );
    } catch (err) {
      expect(err).toMatchSnapshot();
    }
    done();
  });
});
