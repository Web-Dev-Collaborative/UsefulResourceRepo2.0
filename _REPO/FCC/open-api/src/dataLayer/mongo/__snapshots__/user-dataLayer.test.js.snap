// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createUser should throw if accountLinkId is already in db 1`] = `[Error: Account already in use]`;

exports[`deleteUser should refuse deletion of other users 1`] = `[Error: You can delete only your account]`;
